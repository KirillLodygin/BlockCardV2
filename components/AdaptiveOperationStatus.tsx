import React from 'react';
import styled from 'styled-components/macro';

import { AdditionalTypography, ADDITIONAL_TYPOGRAPHY_SIZE } from '@vtb/ui-kit/adaptive/atoms/additionalTypography';
import { MainTypography, MAIN_TYPOGRAPHY_SIZE } from '@vtb/ui-kit/adaptive/atoms/mainTypography';
import { NOTIFICATION_TYPE } from '@vtb/ui-kit/adaptive/molecules/notification';
import { INDENT } from '@vtb/ui-kit/tokens/sizes';
import { TYPOGRAPHY_TAG } from '@vtb/ui-kit/tokens/typography';

import {
  OperationFormStatusAdaptive,
  StatusIconWrapperAdaptive,
  StatusInfoWrapper,
} from '@common/overmind/layouts/common/components/operationForm/AdaptiveOperationFormStatus';
import { ErrorAnimation, SuccessAnimation } from '@common/overmind/layouts/common/layoutParts/NotificationAnimation';

type NotificationType = typeof NOTIFICATION_TYPE.info | typeof NOTIFICATION_TYPE.fail;

type Props = {
  description: string;
  title: string;
  type: NotificationType;
};

const OperationFormStatusAdaptiveWrapper = styled(OperationFormStatusAdaptive)`
  margin-bottom: ${INDENT.m};
`;

const StatusTextAdaptive = styled(MainTypography)`
  margin-bottom: ${INDENT.xxxs};
`;

export const AdaptiveOperationStatus = ({ description, title, type }: Props) => (
  <OperationFormStatusAdaptiveWrapper isError={type === NOTIFICATION_TYPE.fail}>
    <StatusIconWrapperAdaptive>
      {type === NOTIFICATION_TYPE.fail ? <ErrorAnimation /> : <SuccessAnimation />}
    </StatusIconWrapperAdaptive>
    <StatusInfoWrapper>
      <StatusTextAdaptive tag={TYPOGRAPHY_TAG.h2} size={MAIN_TYPOGRAPHY_SIZE.m} autoFocus>
        {title}
      </StatusTextAdaptive>
      <AdditionalTypography size={ADDITIONAL_TYPOGRAPHY_SIZE.m}>{description}</AdditionalTypography>
    </StatusInfoWrapper>
  </OperationFormStatusAdaptiveWrapper>
);
