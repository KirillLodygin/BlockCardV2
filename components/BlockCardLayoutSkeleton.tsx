import React from 'react';
import { Flex, Box } from 'reflexbox/styled-components';

import { AdaptiveCol, AdaptiveRow } from '@vtb/ui-kit/adaptive/atoms/layout';
import { FIELD_WIDTH, INDENT } from '@vtb/ui-kit/tokens/sizes';

import {
  FieldSkeleton,
  DropdownSkeleton,
  TextlineSkeleton,
  LINE_SIZES,
  SKELETON_LINE_WIDTH,
} from '@components/Skeleton/components';
import { ButtonSkeleton } from '@components/Skeleton/components/adaptive/ButtonSkeleton';

const BlockCardLayoutSkeleton = () => (
  <Box mt={INDENT.m}>
    <AdaptiveRow>
      <AdaptiveCol tablet={8}>
        <Flex flexDirection="column">
          <TextlineSkeleton width={SKELETON_LINE_WIDTH.xxl} lineSize={LINE_SIZES.l} />
          <Box mt={INDENT.sm}>
            <FieldSkeleton width={FIELD_WIDTH.auto} />
          </Box>
          <Box mt={INDENT.xxs}>
            <DropdownSkeleton width={FIELD_WIDTH.auto} />
          </Box>
          <Box mt={INDENT.l}>
            <FieldSkeleton width={FIELD_WIDTH.auto} />
          </Box>
          <Box mt={INDENT.xxs}>
            <ButtonSkeleton />
          </Box>
          <Box mt={INDENT.l}>
            <TextlineSkeleton width={SKELETON_LINE_WIDTH.xs} lineSize={LINE_SIZES.l} />
          </Box>
        </Flex>
      </AdaptiveCol>
    </AdaptiveRow>
  </Box>
);

export { BlockCardLayoutSkeleton };
