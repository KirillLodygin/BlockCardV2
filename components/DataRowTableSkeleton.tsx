import React from 'react';
import { Box, Flex } from 'reflexbox/styled-components';

import { INDENT } from '@vtb/ui-kit/tokens/sizes';

import { LINE_SIZES, TextlineSkeleton } from '@components/Skeleton';

export const SkeletonDataRowTableItem = () => (
  <Box mb={INDENT.xxs} width={INDENT.m}>
    <Flex flexGrow={1} mr={INDENT.xxl}>
      <Box ml={INDENT.xxs}>
        <TextlineSkeleton lineSize={LINE_SIZES.xs} />
      </Box>

      <Box ml="auto">
        <TextlineSkeleton lineSize={LINE_SIZES.xs} />
      </Box>
    </Flex>
  </Box>
);

export const DataRowTableSkeleton = () => (
  <Box mt={INDENT.sm} mb={INDENT.m}>
    <SkeletonDataRowTableItem />
    <SkeletonDataRowTableItem />
    <SkeletonDataRowTableItem />
    <SkeletonDataRowTableItem />
  </Box>
);
