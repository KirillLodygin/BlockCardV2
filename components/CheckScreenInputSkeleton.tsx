import React, { FC } from 'react';
import { Box } from 'reflexbox/styled-components';

import { FIELD_WIDTH, INDENT } from '@vtb/ui-kit/tokens/sizes';

import { FieldSkeleton, TextlineSkeleton, LINE_SIZES, SKELETON_LINE_WIDTH } from '@components/Skeleton/components';

export const CheckScreenInputSkeleton: FC = () => (
  <Box mb={INDENT.l}>
    <Box mb={INDENT.sm}>
      <TextlineSkeleton width={SKELETON_LINE_WIDTH.xxl} lineSize={LINE_SIZES.m} />
    </Box>
    <Box maxWidth={FIELD_WIDTH.xl}>
      <FieldSkeleton width={FIELD_WIDTH.auto} />
    </Box>
    <Box mt={INDENT.l}>
      <TextlineSkeleton width={SKELETON_LINE_WIDTH.xxl} lineSize={LINE_SIZES.m} />
    </Box>
  </Box>
);
