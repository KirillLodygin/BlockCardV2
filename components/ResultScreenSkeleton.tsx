import React from 'react';
import { Flex, Box } from 'reflexbox/styled-components';
import styled from 'styled-components/macro';

import { AdaptiveCol } from '@vtb/ui-kit/adaptive/atoms/layout';
import { COLOR_GETTERS } from '@vtb/ui-kit/adaptive/tokens/palette/colorGetters';
import { adaptive } from '@vtb/ui-kit/tokens/screen';
import { INDENT } from '@vtb/ui-kit/tokens/sizes';

import { LINE_SIZES, Skeleton, SKELETON_LINE_WIDTH, TextlineSkeleton } from '@components/Skeleton/components';
import { ButtonSkeleton } from '@components/Skeleton/components/adaptive/ButtonSkeleton';

const NotificationSkeleton = styled.div`
  display: flex;
  flex-direction: row;
  height: 120px;
  background-color: ${COLOR_GETTERS.background.secondary};
  border-radius: ${INDENT.xxxxs};
  margin: 0 0 ${INDENT.m};
  padding: ${INDENT.sm} ${INDENT.s} ${INDENT.m} ${INDENT.xxl};
  box-sizing: border-box;
  gap: ${INDENT.xxxl};

  ${adaptive.maxWidth.mobile} {
    align-items: center;
    flex-direction: column;
    height: 100%;
    padding: ${INDENT.s};
    width: 100%;
  }
`;

const AdaptiveIndentContainer = styled(Flex)`
  flex-direction: row;
  gap: ${INDENT.m};

  ${adaptive.maxWidth.mobile} {
    flex-direction: column;
    width: 100%;
  }
`;

const ResultScreenSkeleton = () => (
  <>
    <NotificationSkeleton>
      <Box>
        <Skeleton width={64} height={64} radius={32} />
      </Box>
      <AdaptiveCol tablet={8}>
        <TextlineSkeleton width={SKELETON_LINE_WIDTH.xl} lineSize={LINE_SIZES.m} />
        <TextlineSkeleton width={SKELETON_LINE_WIDTH.m} lineSize={LINE_SIZES.l} />
      </AdaptiveCol>
    </NotificationSkeleton>
    <AdaptiveIndentContainer>
      <ButtonSkeleton buttonBorderRadius={4} />
      <ButtonSkeleton buttonBorderRadius={4} />
    </AdaptiveIndentContainer>
  </>
);

export { ResultScreenSkeleton };
