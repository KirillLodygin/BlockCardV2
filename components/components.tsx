import React, { useEffect, useRef, useState } from 'react';
import { Box, Flex } from 'reflexbox/styled-components';
import styled from 'styled-components/macro';

import { AdditionalTypography, ADDITIONAL_TYPOGRAPHY_SIZE } from '@vtb/ui-kit/adaptive/atoms/additionalTypography';
import { ButtonText } from '@vtb/ui-kit/adaptive/atoms/button';
import { Icon, ICON_SIZE } from '@vtb/ui-kit/adaptive/atoms/icon';
import { MAIN_TYPOGRAPHY_SIZE, MainTypography } from '@vtb/ui-kit/adaptive/atoms/mainTypography';
import { Notification as Notice, NOTIFICATION_TYPE } from '@vtb/ui-kit/adaptive/molecules/notification';
import { COLOR_GETTERS } from '@vtb/ui-kit/adaptive/tokens/palette/colorGetters';
import { CardPending, ChevronLeft } from '@vtb/ui-kit/assets';
import { INDENT } from '@vtb/ui-kit/tokens/sizes';
import { TYPOGRAPHY_WEIGHT } from '@vtb/ui-kit/tokens/typography';

import { ReactComponent as CardBlockErrorDark } from '@assets/cards/cardBlockErrorDark.svg';
import { ReactComponent as CardBlockErrorLight } from '@assets/cards/cardBlockErrorLight.svg';
import { ReactComponent as CardBlockSuccessDark } from '@assets/cards/cardBlockSuccessDark.svg';
import { ReactComponent as CardBlockSuccessLight } from '@assets/cards/cardBlockSuccessLight.svg';
import { useThemedContent } from '@common/appTheme';
import { OPERATIONS_CONSTANTS } from '@common/overmind/constants/base';
import { DEFAULT_ERROR_MSG } from '@common/overmind/constants/messages';
import { OperationFormChild } from '@common/overmind/layouts/common/components/operationForm';
import {
  OperationFormStatus,
  StatusIconContainer,
  StatusText,
} from '@common/overmind/layouts/common/components/operationForm/OperationFormStatus';
import { ELEMENT_TYPES } from '@common/overmind/layouts/common/constants';
import { cutBBCodesAndTags, parseBBCode } from '@common/parse-bb-code';
import { useSetNavBarTitle } from '@modules/NavBar';

// TODO-gordeev: использовать более подходящие типы после того как они будут сформированы в овермайнде
// TODO-svetlov: перетащить все компоненты в свои файлы и переместить в папку с компонентами

type TitleProps = {
  titleText?: string;
};

export const Title = ({ titleText }: TitleProps) => (
  <Box mb={INDENT.sm}>
    <MainTypography autoFocus>{parseBBCode(titleText)}</MainTypography>
  </Box>
);

const NotificationText = styled(AdditionalTypography).attrs({
  size: ADDITIONAL_TYPOGRAPHY_SIZE.m,
})`
  color: ${COLOR_GETTERS.text.secondary};
`;

type NotificationProps = {
  hasErrorNotification: boolean;
  isLockingSuccess: boolean;
  statusText: string | undefined;
  details: React.Component;
};

export const Notification = ({ statusText, hasErrorNotification, isLockingSuccess, details }: NotificationProps) => {
  const illustrations = useThemedContent(
    {
      success: <CardBlockSuccessLight />,
      error: <CardBlockErrorLight />,
    },
    {
      success: <CardBlockSuccessDark />,
      error: <CardBlockErrorDark />,
    },
  );

  return (
    <Box mb={INDENT.l}>
      <OperationFormStatus>
        <StatusIconContainer>
          {isLockingSuccess && illustrations.success}
          {hasErrorNotification && illustrations.error}
        </StatusIconContainer>
        <StatusText autoFocus>{statusText}</StatusText>

        <NotificationText>{details}</NotificationText>
      </OperationFormStatus>
    </Box>
  );
};

type ButtonsBlockProps = {
  buttons: ReadonlyArray<any>;
};

export const ButtonsBlock = ({ buttons }: ButtonsBlockProps) => (
  <Box mt={INDENT.m}>
    {buttons.map((button) =>
      button.props.name === OPERATIONS_CONSTANTS.commandNames.CANCEL ? (
        <Box mt={INDENT.l} key={button.key}>
          {button}
        </Box>
      ) : (
        <Box key={button.key}>{button}</Box>
      ),
    )}
  </Box>
);

type ElementsBlockProps = {
  elements: ReadonlyArray<any>;
};

export const ElementsBlock = ({ elements }: ElementsBlockProps) => (
  <Box mb={INDENT.xs} flexDirection="column" key="rootGroup">
    {elements.map((field, i) => (
      <OperationFormChild type={ELEMENT_TYPES.FIELD} key={i}>
        {field}
      </OperationFormChild>
    ))}
  </Box>
);

export const BlockTitle = ({ titleText }: TitleProps) => {
  const TitleRef = useRef(null);
  useSetNavBarTitle(titleText, TitleRef);

  if (titleText) {
    return (
      <Box mb={INDENT.m}>
        <MainTypography size={MAIN_TYPOGRAPHY_SIZE.m} tag="h3" ref={TitleRef} autoFocus>
          {titleText}
        </MainTypography>
      </Box>
    );
  }

  return null;
};

export const DefaultErrorBlock = () => (
  <Box mt={INDENT.m}>
    <Notice text={DEFAULT_ERROR_MSG} type={NOTIFICATION_TYPE.fail} />
  </Box>
);

const StyledAdditionalTypography = styled(AdditionalTypography)`
  color: ${COLOR_GETTERS.text.primary};
`;

type CodeRequestTimerProps = {
  isTimerRunning: boolean;
  setIsTimerRunning: (value: boolean) => void;
  text: string;
  timeout: number;
};

export const CodeRequestTimer = ({ isTimerRunning, setIsTimerRunning, text, timeout }: CodeRequestTimerProps) => {
  const [seconds, setSeconds] = useState(timeout);

  useEffect(() => {
    const timerId = setInterval(() => {
      setSeconds(seconds - 1);
    }, 1000);

    if (seconds === 0) {
      clearInterval(timerId);
      setIsTimerRunning(false);
    }

    return () => clearInterval(timerId);
  }, [seconds, setIsTimerRunning]);

  useEffect(() => {
    if (isTimerRunning) {
      setSeconds(timeout);
    }
  }, [isTimerRunning, timeout]);

  return (
    <Box mt={INDENT.xs}>
      <Flex>
        {isTimerRunning && <Icon type={CardPending} size={ICON_SIZE.s} />}
        <StyledAdditionalTypography weight={TYPOGRAPHY_WEIGHT.bold} role="timer">
          <Box as="span" ml={INDENT.xxxxs}>
            {cutBBCodesAndTags(text, seconds.toString().padStart(2, '0'))}
          </Box>
        </StyledAdditionalTypography>
      </Flex>
    </Box>
  );
};

type BackButtonBlockProps = {
  children: string;
  'data-test-id': string;
  disabled?: boolean;
  onClick: () => void;
};

export const BackButtonBlock = ({ children, ...props }: BackButtonBlockProps) => (
  <ButtonText size="medium" icon={ChevronLeft} {...props}>
    {children}
  </ButtonText>
);
