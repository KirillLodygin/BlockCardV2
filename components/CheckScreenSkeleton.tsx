import React from 'react';
import { Flex, Box } from 'reflexbox/styled-components';

import { AdaptiveCol, AdaptiveRow } from '@vtb/ui-kit/adaptive/atoms/layout';
import { FIELD_WIDTH, INDENT } from '@vtb/ui-kit/tokens/sizes';

import { LINE_SIZES, SKELETON_LINE_WIDTH, TextlineSkeleton } from '@components/Skeleton/components';
import { ButtonSkeleton } from '@components/Skeleton/components/adaptive/ButtonSkeleton';

import { CheckScreenInputSkeleton } from './CheckScreenInputSkeleton';
import { DataRowTableSkeleton } from './DataRowTableSkeleton';

const CheckScreenSkeleton = () => (
  <Box mt={INDENT.m}>
    <AdaptiveRow>
      <AdaptiveCol tablet={8}>
        <TextlineSkeleton width={SKELETON_LINE_WIDTH.xxl} lineSize={LINE_SIZES.l} />
        <DataRowTableSkeleton />
        <CheckScreenInputSkeleton />
        <Flex alignItems="center" flexDirection="row" maxWidth={FIELD_WIDTH.xl} mb={INDENT.m}>
          <ButtonSkeleton buttonBorderRadius={4} />
          <Box ml={INDENT.m}>
            <ButtonSkeleton buttonBorderRadius={4} />
          </Box>
        </Flex>
        <Box mt={INDENT.l}>
          <TextlineSkeleton width={SKELETON_LINE_WIDTH.xs} lineSize={LINE_SIZES.l} />
        </Box>
      </AdaptiveCol>
    </AdaptiveRow>
  </Box>
);

export { CheckScreenSkeleton };
