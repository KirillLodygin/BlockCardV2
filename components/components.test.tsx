import { shallow } from 'enzyme';
import React from 'react';

import { ButtonText } from '@vtb/ui-kit/adaptive/atoms/button';

import { BackButtonBlock, CodeRequestTimer } from './components';

const mockOnClick = jest.fn();

describe('components', () => {
  describe('BackButtonBlock', () => {
    it('should calls backButtonAction', () => {
      const component = shallow(
        <BackButtonBlock onClick={mockOnClick} data-test-id="qwe">
          foo
        </BackButtonBlock>,
      );

      component.find(ButtonText).simulate('click');

      expect(mockOnClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('CodeRequestTimer', () => {
    it('should render with isTimerRunning=true', () => {
      const component = shallow(
        <CodeRequestTimer isTimerRunning setIsTimerRunning={jest.fn()} text="foo" timeout={11} />,
      );

      expect(component).toMatchSnapshot();
    });

    it('should render with isTimerRunning=false', () => {
      const component = shallow(
        <CodeRequestTimer isTimerRunning={false} setIsTimerRunning={jest.fn()} text="foo" timeout={11} />,
      );

      expect(component).toMatchSnapshot();
    });
  });
});
