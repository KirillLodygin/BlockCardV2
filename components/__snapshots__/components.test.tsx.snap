// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components CodeRequestTimer should render with isTimerRunning=false 1`] = `
<styled.div
  mt="16px"
>
  <Styled(styled.div)>
    <components__StyledAdditionalTypography
      role="timer"
      weight="bold"
    >
      <styled.div
        as="span"
        ml="4px"
      >
        foo
      </styled.div>
    </components__StyledAdditionalTypography>
  </Styled(styled.div)>
</styled.div>
`;

exports[`components CodeRequestTimer should render with isTimerRunning=true 1`] = `
<styled.div
  mt="16px"
>
  <Styled(styled.div)>
    <Icon
      size={24}
      type="cardPending.svg"
    />
    <components__StyledAdditionalTypography
      role="timer"
      weight="bold"
    >
      <styled.div
        as="span"
        ml="4px"
      >
        foo
      </styled.div>
    </components__StyledAdditionalTypography>
  </Styled(styled.div)>
</styled.div>
`;
